### PR TITLE
RFC FS-1005 - Underscore literals

### PR DESCRIPTION
**Pull request by AviAvni**
I implemented the feature according to the [RFC](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1005-underscores-in-numeric-literals.md)
I samples from RFC work properly please review before I added the tests

### CODE REVIEW

**Comment by forki** [comment]
Can you please start adding the tests? Would make it much easier for me to review.

**Comment by AviAvni** [comment]
OK working on it thanks

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 1]
```diff
@@ -4144,8 +4144,8 @@ namespace Microsoft.FSharp.Core
         let inline ParseUInt16 (s:string)     = (# "conv.ovf.u2" (ParseUInt32 s) : uint16 #)
         let inline ParseIntPtr (s:string)  = (# "conv.ovf.i"  (ParseInt64 s)  : nativeint #)
         let inline ParseUIntPtr (s:string) = (# "conv.ovf.u"  (ParseInt64 s)  : unativeint #)
-        let inline ParseDouble (s:string)   = Double.Parse(s,NumberStyles.Float, CultureInfo.InvariantCulture)
-        let inline ParseSingle (s:string) = Single.Parse(s,NumberStyles.Float, CultureInfo.InvariantCulture)
+        let inline ParseDouble (s:string)   = Double.Parse(s.Replace("_", ""),NumberStyles.Float, CultureInfo.InvariantCulture)
```
Comment (forki) [inline src/fsharp/FSharp.Core/prim-types.fs]: Why is there no trim in this place? Ideas?

File: src/fsharp/lex.fsl [hunk 2]
```diff
@@ -83,6 +83,7 @@ let parseOctalUInt64 (s:string) p l =
     parse p 0UL
 
 let parseInt32 (s:string) = 
+    let s = s.Replace("_", "")
```
Comment (forki) [inline src/fsharp/lex.fsl]: Can we make this a separate (inlined) helper function?

**Comment by AviAvni** [comment]
I added the positive test
for the negative it's look like I need vs15 to run them so i'm moving to the virtual machine

File: src/fsharp/FSharp.Core/prim-types.fs (hunk 1 above)
Comment (AviAvni) [inline src/fsharp/FSharp.Core/prim-types.fs]: Why in the other places there is a trim?
As I see in the parser there are no leading or trailing whitespaces in numeric literals

File: src/fsharp/lex.fsl (hunk 2 above)
Comment (AviAvni) [inline src/fsharp/lex.fsl]: Done

File: tests/fsharpqa/Source/Conformance/BasicGrammarElements/Constants/BasicConstants.fs [hunk 3]
```diff
@@ -40,4 +40,19 @@ let boolConst2 = false
...
+let hexWords = 0xCAFE_BABE
+let maxLong = 0x7fff_ffff_ffff_ffffL
+let nybbles = 0b0010_0101
+let bytes = 0b11010010_01101001_10010100_10010010
+let x2 = 5_2
+let x4 = 5_______2
+let x7 = 0x5_2
+let x9 = 0_52
+let x10 = 05_2
+let x14 = 0o5_2
```
Comment (forki) [inline tests/fsharpqa/Source/Conformance/BasicGrammarElements/Constants/BasicConstants.fs]: I think we need tests that show that the values are actually the expected constants.

**Comment by forki** [comment]
added few comments. apart from that:

underscore

File: tests/fsharpqa/Source/Conformance/BasicGrammarElements/Constants/BasicConstants.fs (hunk 3 above)
Comment (AviAvni) [inline tests/fsharpqa/Source/Conformance/BasicGrammarElements/Constants/BasicConstants.fs]: Right sorry done

**Comment by forki** [comment]
somehow some tests are broken

    [<Test>]
    member this.ParseStringViaConversionOps() =
        let s : string = null
        CheckThrowsArgumentNullException2 "sbyte" (fun () -> sbyte s |> ignore)
        CheckThrowsArgumentNullException2 "byte" (fun () -> byte s |> ignore)
        CheckThrowsArgumentNullException2 "int16" (fun () -> int16 s |> ignore)
        CheckThrowsArgumentNullException2 "uint16 " (fun () -> uint16 s |> ignore)
        CheckThrowsArgumentNullException2 "int" (fun () -> int s |> ignore)
        CheckThrowsArgumentNullException2 "int32" (fun () -> int32 s |> ignore)
        CheckThrowsArgumentNullException2 "uint32" (fun () -> uint32 s |> ignore)
        CheckThrowsArgumentNullException2  "int64" (fun () -> int64 s |> ignore)
        CheckThrowsArgumentNullException2 "uint64" (fun () -> uint64 s |> ignore)
        CheckThrowsArgumentNullException2 "float32" (fun () -> float32 s |> ignore)
        CheckThrowsArgumentNullException2 "float" (fun () -> float s |> ignore)
        CheckThrowsArgumentNullException2 "decimal" (fun () -> decimal s |> ignore)

looks like it's related to parsing/converting nulls

**Comment by AviAvni** [comment]
Tanks you are great i'll check it tomorrow night

File: src/fsharp/FSharp.Core/prim-types.fs [hunk 4]
```diff
@@ -2504,10 +2504,13 @@ namespace Microsoft.FSharp.Core
             let rec parse n acc = if n < l then parse (n+1) (acc *.. 2UL +.. (match s.Chars(n) with '0' -> 0UL | '1' -> 1UL | _ -> formatError())) else acc in          
             parse p 0UL
 
+        let inline removeUnderscores (s:string) =
+            s.Replace("_", "")
```
Comment (forki) [inline src/fsharp/FSharp.Core/prim-types.fs]: I think this needs a null check to fix the tests.

    match s with
    | null -> s
    | _ -> s.Replace("_", "")

Same with other instance of this method in lex.fsl

**Comment by AviAvni** [comment]
@dotnet-bot test this please

**Comment by AviAvni** [comment]
@dsyme what do you think about the implementation?
the errors message for invalid literals in the RFC are the errors the compile need to report or the
regular error for invalid numeric literals is ok?
I have a problem to add test to the errors it's tell me the error no match and then print the same error

File: src/fsharp/lex.fsl [hunk 5]
```diff
@@ -82,7 +82,13 @@ let parseOctalUInt64 (s:string) p l =
     let rec parse n acc = if n < l then parse (n+1) (acc * 8UL + (let c = s.[n] in if c >= '0' && c <= '7' then Convert.ToUInt64 c - Convert.ToUInt64 '0' else formatError())) else acc 
     parse p 0UL
 
+let inline removeUnderscores (s:string) =
```
Comment (dsyme) [inline src/fsharp/lex.fsl]: There's no need to mark this "inline" though it's prety harmless

File: src/fsharp/lex.fsl [hunk 6]
```diff
@@ -175,11 +181,12 @@ let anychar = [^'\n''\r']
 let anystring = anychar*
 let op_char = '!'|'$'|'%'|'&'|'*'|'+'|'-'|'.'|'/'|'<'|'='|'>'|'?'|'@'|'^'|'|'|'~'|':'
 let ignored_op_char = '.' | '$' | '?'
+let seperator = '_'
```
Comment (dsyme) [inline src/fsharp/lex.fsl]: spelling is "separator" though it's no big deal :)

**Comment by dsyme** [comment]
LGTM - I think this is ready to go.

**Comment by AviAvni** [comment]
@dsyme thanks I fixed the typo and inline issues

**Comment by khyperia** [comment]
@AviAvni You seem to be missing a test case for the following bit of the spec:

> You cannot place underscores in the following places:
> * Prior to an F or L or other suffix

and I noticed that the implementation grammar seems to allow it (`1.0_F`), although I can't test it myself at the moment. Would you mind adding a test case for that?

**Comment by khyperia** [comment]
Also, the spec doesn't seem to make any stance on the validity of underscores around the exponent notation (`2.1_e2F`, `2.1e_2F`) - I think the grammar currently allows it, although it seems consistent to disallow it (for example, C#7's similar feature disallows underscores around an exponent). In any case, we probably want to add a test asserting one behavior or another.

**Comment by AviAvni** [comment]
@khyperia thanks I opened new pr fixed the float issue and add tests for errors https://github.com/Microsoft/visualfsharp/pull/1256

File: tests/fsharpqa/Source/Conformance/BasicGrammarElements/Constants/BasicConstants.fs [hunk 7]
```diff
@@ -40,4 +40,60 @@ let boolConst2 = false
...
 
+let creditCardNumber = 1234_5678_9012_3456L
+if creditCardNumber <> 1234567890123456L then
+    failwith "Wrong parsing"
+
+let socialSecurityNumber = 999_99_9999L
+if socialSecurityNumber <> 999999999L then
+    failwith "Wrong parsing"
+
+let pi =    3.14_15F
```
Comment (jcouv) [inline tests/fsharpqa/Source/Conformance/BasicGrammarElements/Constants/BasicConstants.fs]: spacing is odd

Comment (forki) [inline tests/fsharpqa/Source/Conformance/BasicGrammarElements/Constants/BasicConstants.fs]: I think that's by design to test more interesting cases.

**Pull request merged**